mobius-core/src/main/java/com/spotify/mobius/EventSourceConnectable.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * -\-\-
+ * Mobius
+ * --
+ * Copyright (c) 2017-2018 Spotify AB
+ * --
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * -/-/-
+ */
+package com.spotify.mobius;
+
+import com.spotify.mobius.disposables.Disposable;
+import com.spotify.mobius.functions.Consumer;
+import javax.annotation.Nonnull;
+
+class EventSourceConnectable<M, E> implements Connectable<M, E> {
+
+  public static <M, E> Connectable<M, E> create(EventSource<E> source) {
+    return new EventSourceConnectable<>(source);
+  }
+
+  private final EventSource<E> eventSource;
+
+  private EventSourceConnectable(EventSource<E> source) {
+    eventSource = source;
+  }
+
+  @Nonnull
+  @Override
+  public Connection<M> connect(final Consumer<E> output) throws ConnectionLimitExceededException {
+    return new Connection<M>() {
+      private Disposable disposable;
+
+      @Override
+      public void accept(M value) {
+        if (disposable != null) {
+          return;
+        }
+        disposable = eventSource.subscribe(output);
```

Comment:
This is weird to me. It means "only subscribe to event sources after `init` has been executed and the first model emitted", right? I don't think that's a constraint that we should enforce on users of Mobius. If it is common that users want that, then that behaviour should be enabled. But if it is the case that users want to subscribe earlier, and manage concurrent event emission some other way, it doesn't feel right to make that impossible. 

It's a behaviour change, too, IIUC. Before, `MobiusLoop.startFrom` was synchronous with `EventSource.subscribe`, and it happened before `init`. Now, when `subscribe` happens always depends on how you've configured event processing. 

Is what we need perhaps to add the ability to add `EventSource` or `Connectable<M, E>` to the builder, as separate things rather than combine them? I don't think a 'change' is necessarily backwards compatible (at the API level, this one is, but not behaviourally), so perhaps 'add' is a better choice?

Reply:
1. Currently, a loop subscribes immediately to an event source, but queues up all the events it generates until init has completed running. This is the same as subscribing to the event source after init has completed running. In fact, nothing about the loop's API indicates when the subscription to the event source is actually done. 
2. I'll also point out that this change passes all the tests, which means the behavior is still intact.

I could change it so that this thing subscribes to the event source the moment its `connect()` method is called. This would be the exact same thing as the loop previously did. But I don't think this would be the correct behavior, especially that we have tests guarding against that exact race condition, and I'll argue that we only have this problem because we didn't have the correct abstraction for event sources to begin with. Which makes me realize now that this test is useless with the change that I've made and it needs to be changed to a Connectable that is eager, not just an EventSource. I'll modify that test.

Finally, an `EventSource` should not be coupled with when a loop is going to subscribe to it. It shouldn't even be aware of the lifecycle of a loop. I'll argue that subscribing to the event source in the loop's constructor, then queuing all its events until the loop has finished running init, is the same as deferring the subscription until the loop has initialized, since the subscription should generate the same events regardless of when it happens.

Reply:
Also you say "But if it is the case that users want to subscribe earlier, and manage concurrent event emission some other way, it doesn't feel right to make that impossible.". I would say that users can still do that by creating an `EventSource` that subscribes to stuff in its constructor. 

This does sound like a pretty odd thing to me though. `EventSource<E>` is our isomorphism to `Observable<E>`. An `EventSource<E>` can only possibly start sending events to the loop when its `subscribe()` method is invoked by the loop. This is the same as `Observable.subscribe`. Therefore, it is correct to say that `EventSource<E>` is lazy and potentially cold, unless the implementation multicasts the items that it emits.. Now that being said, it can absolutely have a dependency on another observable that is hot. But this is still an implementation detail of the `EventSource` itself, and cannot really be built with assumptions about when a loop subscribes to the event source.

Reply:
> 1. Currently, a loop subscribes immediately to an event source, but queues up all the events it generates until init has completed running. This is the same as subscribing to the event source after init has completed running. In fact, nothing about the loop's API indicates when the subscription to the event source is actually done.

I don't think this is correct. Let's say the event source generates events 'continuously', without being aware of the lifecycle of the loop. It generates events A, B, and C, in order. Let's suppose in some execution, A is generated before 'init' completes, and 'B' after. Before this merge, A could be queued and executed, whereas after it, it cannot. It's true that there's nothing in the API that describes this, and I don't think any code should depend on it. But AFAICT, it does represent a change in behaviour.

> 2. I'll also point out that this change passes all the tests, which means the behavior is still intact.

Only insofar as we have written tests that cover all kinds of behaviour. :) I don't think we have.

Another behaviour change that I wasn't explicit about is which thread invokes `subscribe`. Again, something that isn't defined in the API, but that could potentially break some users.

I don't understand what you mean by saying "we didn't have the right abstraction for event sources"? To me, it was a good thing that event sources are different from Elm's Subscriptions, because it means they can be simpler. You can still make an `EventSource` that observes the model emissions from the loop, as you show with this PR. I like that decoupling. What about event sources is the wrong abstraction?

Reply:
The reason I don't think `EventSource` is the correct abstraction is because it is overly simplified, and does not allow for the more advanced case. More specifically, this abstraction does not work well if you're trying to compose domains. Especially if the subdomain does not necessarily begin with its parent. So while `EventSource` for a single domain might work ok, it instantly fails when you try to compose domains. Furthermore, right now if I want to subscribe to different event sources based on the state that I am in, I would have to create a special effect that I dispatch every time the model changes, and I would need to pack the model into that effect, and have the effect handler do its thing. It solves the case, but is more error prone since it's possible to forget dispatching that effect. Furthermore, I would have to repeat this every time I needed this mechanism.

Whereas treating event sources as a Connectable actually allows us to do both; subscribe only once, or do conditional subscriptions.

Reply:
I experimented with the `EventSource` that observes the loop, but I found that it would lead to cyclical dependencies and the only way to work around it is to use something like RxJava's `BehaviorSubject`, and things just got nastier from that point onwards.

I do agree that this changes the thread which subscriptions occur, but again, this is not indicated in the API at all, and it makes these assumptions about things when it shouldn't.

If I haven't managed to convince you yet with my arguments above, I'd be willing to change this so that it subscribes in the `connect` method, and not wait for the first model to come in. This would be doing the exact thing that the loop is doing right now. But I really do think that a deferred vs eager subscription will be the same, but the deferred one is more correct.

Reply:
I'm thinking of the `EventSource` as  a wrapper around something else that may fire things 'at any point in time'. I would say the event source itself can only emit events after a `subscribe` call, but the underlying events can happen at any time (messages pushed from a server, or whatever) irrespective of the lifecycle of the wrapper. So the change isn't really in the `EventSource`, but in when its events get observable by the Mobius loop. It's probably not super-important, but it does represent a change in behaviour.

I'm not opposed to adding `Connectable` as an alternative to `EventSource`. AFAIU, that would solve your use case without changing behaviours or removing the 'simple' case of something external that fires events without wanting to observe the model. But I don't see the value in combining the two into one thing - can you describe why that is useful or necessary?